fastify integration (#174)
dmno-dev/dmno: packages/docs-site/src/content/docs/docs/integrations/fastify.mdx

---
title: Fastify
description: Use DMNO to manage your Fastify app's environment variables
npmPackage: "@dmno/fastify-integration"
---

import { Tabs, TabItem } from '@astrojs/starlight/components'
import TabbedCode from '@/components/TabbedCode.astro';
import BugReportLink from '@/components/BugReportLink.astro';

> Fastify does not provide any built-in handling of env vars, although there is a [fastify-env](https://github.com/fastify/fastify-env) plugin, which internally uses dotenv files with json schema and ajv for validations. Whether you use that plugin, or have wired up something else yourself, we think using DMNO instead will provide many benefits.

Reviewer: ```suggestion
Fastify does not provide any built-in handling of env vars, although there is a [fastify-env](https://github.com/fastify/fastify-env) plugin, which internally uses dotenv files with json schema and ajv for validations. Whether you use that plugin, or have wired up something else yourself, we think using DMNO instead is worth the additional benefits.
```

:::note[Compatibility - TypeScript + ESM]

> This plugin works regardless of if you are using JavaScript or TypeScript, but currently it only supports ESM projects.

Reviewer: ```suggestion
This plugin works if you are using JavaScript or TypeScript, but currently it only supports ESM projects.
```

<TabbedCode packageName="@dmno/fastify-integration dmno" />

The rest of the Fastify setup looks slightly different depending on if you areusing the [fastify-cli](https://github.com/fastify/fastify-cli) or not. Select your situation using the tabs below:

#### Initialize + register `dmnoFastifyPlugin`
<Tabs syncKey='fastifyType'>
  <TabItem label="Direct usage">
    Wherever you initialize your `fastify` instance and register plugins, import and register our `dmnoFastifyPlugin`:

    ```diff lang="js"
    import Fastify from 'fastify';
    +import { dmnoFastifyPlugin } from '@dmno/fastify-integration';

    const fastify = Fastify({ /* ... */ });
    +fastify.register(dmnoFastifyPlugin);
    fastify.register(someOtherPlugin);
    ```
  </TabItem>
  <TabItem label="fastify-cli">
    Create a new file in the plugins directory to initialize the plugin:

    ```diff lang="js" title="src/plugins/dmno.js"
    import { dmnoFastifyPlugin } from "@dmno/fastify-integration";
    export default dmnoFastifyPlugin;
    ```
  </TabItem>
</Tabs>

#### Adjust your package.json scripts

> Unlike some of our front-end integrations where we hook into an existing build process, we must run our dev and boot commands via [`dmno run`](/docs/reference/cli/run/). You'll want to adjust your `package.json` scripts accordingly. Your existing scripts may not match exactly, but that's ok.

Reviewer: ```suggestion
In this case, we must run our `dev` and `start` commands via [`dmno run`](/docs/reference/cli/run/). You'll want to adjust your `package.json` scripts accordingly. Your existing scripts may not match exactly, but that's ok. Just note that if you want live reload you need to include the `-w` flag. 
```
